purchases/src/main/kotlin/com/revenuecat/purchases/Backend.kt

```diff
@@ -280,34 +296,59 @@ internal class Backend(
         attributes: Map<String, SubscriberAttribute>,
         appUserID: String,
         onSuccessHandler: () -> Unit,
-        onErrorHandler: (PurchasesError, syncedSuccessfully: Boolean) -> Unit
+        onErrorHandler: (PurchasesError, didBackendGetAttributes: Boolean, attributeErrors: List<SubscriberAttributeError>) -> Unit
     ) {
         enqueue(object : Dispatcher.AsyncCall() {
             override fun call(): HTTPClient.Result {
-                val attributesMap = attributes.map { (key, subscriberAttribute) ->
-                    key to subscriberAttribute.toBackendMap()
-                }.toMap()
                 return httpClient.performRequest(
                     "/subscribers/" + encode(appUserID) + "/attributes",
-                    mapOf("attributes" to attributesMap),
+                    mapOf("attributes" to attributes.toBackendMap()),
                     authenticationHeaders
                 )
             }
 
             override fun onError(error: PurchasesError) {
-                onErrorHandler(error, false)
+                onErrorHandler(error, false, emptyList())
             }
 
             override fun onCompletion(result: HTTPClient.Result) {
                 if (result.isSuccessful()) {
                     onSuccessHandler()
                 } else {
-                    onErrorHandler(result.toPurchasesError(), result.responseCode < HTTP_SERVER_ERROR_CODE)
+                    val error = result.toPurchasesError()
+                    var attributeErrors: List<SubscriberAttributeError> = emptyList()
+                    if (error.code == PurchasesErrorCode.InvalidSubscriberAttributesError) {
+                        attributeErrors = result.body!!.getAttributeErrors()
+                    }
+                    onErrorHandler(error, result.responseCode < HTTP_SERVER_ERROR_CODE, attributeErrors)
                 }
             }
         })
     }
 
+    private fun JSONObject.getAttributeErrors(): List<SubscriberAttributeError> {
+        val attributesErrorsList = mutableListOf<SubscriberAttributeError>()
+        val attributeErrorsJSONObject =
+            if (has(ATTRIBUTES_ERROR_RESPONSE_KEY)) getJSONObject(ATTRIBUTES_ERROR_RESPONSE_KEY) else this
+        if (attributeErrorsJSONObject.has(ATTRIBUTE_ERRORS)) {
+            getJSONArray(ATTRIBUTE_ERRORS).let { jsonArray ->
+                for (i in 0 until jsonArray.length()) {
+                    with(jsonArray.getJSONObject(i)) {
+                        if (has("key_name") && has("message")) {
+                            attributesErrorsList.add(
+                                SubscriberAttributeError(
+                                    getString("key_name"),
+                                    getString("message")
+                                )
+                            )
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
this nesting is pretty deep. 
maybe functional programming can help? 
something along the lines of 
(pseudo)
```kotlin
getJSONArray(ATTRIBUTE_ERRORS).let { jsonArray ->
    attributesErrorsList =
            0..jsonArray.length
                    .map { index -> jsonArray.getJSONObject(index) }
                    .filter { has("key_name") && has("message") }
                    .map {
                        SubscriberAttributeError(getString("key_name"), getString("message")
                    }
```